prow/plugins/hold/hold.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package hold contains a plugin which will allow users to label their
+// own pull requests as not ready or ready for merge. The submit queue
+// will honor the label to ensure pull requests do not merge when it is
+// applied.
+package hold
+
+import (
+	"fmt"
+	"regexp"
+
+	"github.com/Sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/plugins"
+)
+
+const pluginName = "hold"
+
+var (
+	label         = "do-not-merge/hold"
+	labelRe       = regexp.MustCompile(`(?mi)^/hold\s*$`)
+	labelCancelRe = regexp.MustCompile(`(?mi)^/hold cancel\s*$`)
+)
+
+type event struct {
+	org      string
+	repo     string
+	number   int
+	body     string
+	htmlurl  string
+	hasLabel func() (bool, error)
+}
+
+func init() {
+	plugins.RegisterIssueCommentHandler(pluginName, handleIssueComment)
+	plugins.RegisterReviewEventHandler(pluginName, handleReview)
+	plugins.RegisterReviewCommentEventHandler(pluginName, handleReviewComment)
+}
+
+type githubClient interface {
+	AddLabel(owner, repo string, number int, label string) error
+	RemoveLabel(owner, repo string, number int, label string) error
+	GetIssueLabels(org, repo string, number int) ([]github.Label, error)
+}
+
+func handleIssueComment(pc plugins.PluginClient, ic github.IssueCommentEvent) error {
+	// Only consider open PRs.
+	if !ic.Issue.IsPullRequest() || ic.Issue.State != "open" || ic.Action != github.IssueCommentActionCreated {
+		return nil
+	}
+
+	e := &event{
+		org:    ic.Repo.Owner.Login,
+		repo:   ic.Repo.Name,
+		number: ic.Issue.Number,
+		body:   ic.Comment.Body,
+		hasLabel: func() (bool, error) {
+			return ic.Issue.HasLabel(label), nil
+		},
+		htmlurl: ic.Comment.HTMLURL,
+	}
+	return handle(pc.GitHubClient, pc.Logger, e)
+}
+
+func handleReview(pc plugins.PluginClient, re github.ReviewEvent) error {
+	if re.Action != github.ReviewActionSubmitted {
+		return nil
+	}
+
+	var (
```

Comment:
nit: this block isn't really necessary now, just put the org and such in the event.

Reply:
They're used both in the event and in the `hasLabel` function for the event, did you mean I should just write them out long-form in both situations?

Reply:
Oh, I missed the `hasLabel` usage. Up to you, then. I don't mind either way.

Reply:
Less pressure on that overworked CI bot if we don't push another change :)